.github/workflows/build.yml

```diff
@@ -5,21 +5,80 @@ on:
     branches:
       - master
 jobs:
-  check-code-formatting-c-cpp:
-    name: Check code formatting (C/C++)
+  analyze-changes:
+    name: Analyze changes
+    runs-on: ubuntu-latest
+    outputs:
+      run-build: ${{steps.change-analysis.outputs.run-build}}
+      run-format-check: ${{steps.change-analysis.outputs.run-format-check}}
+      code-changes: ${{steps.change-analysis.outputs.code-changes}}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2.3.4
+        with:
+          fetch-depth: 0
+      - name: Analyze changes
+        id: change-analysis
+        shell: bash
+        run: |
+          if [ '${{github.ref}}' == 'refs/heads/master' ]; then
+            build_changes='true'
+            format_changes='true'
+            code_changes=$(
+              find . -type f \
+              -iregex '^.*\.\(c\|cpp\|h\|hpp\|inl\)$' -and -not \
+              -iregex '^\.\/externals.*$'
+            )
+          else
+            build_changes=$(
+              git diff --name-only origin/master HEAD -- \
+              ':!*.md' ':!*.asciidoc' ':!.github/CODEOWNERS' ':!externals/*/*' \
```

Comment:
I have some ideas for making this list be data driven eventually. For now it's just hard-coded and we can adjust as needed.